olingo/src/main/java/org/teiid/olingo/service/LocalClient.java

```diff
@@ -236,12 +257,66 @@ public void executeSQL(Query query, List<SQLParameter> parameters,
             }
             getCount = false; // the URL might have $count=true, but ignore it.
         }
+        
+        int count = 0;
+        int expectedEnd = 0;
+        
+        if (!getCount && cache) {
+            //to prevent long initial requests, we want to
+            //only work over windows of at most 32 pages
+            int resultWindow = pageSize<<5;
+            int windows = 0;
```

Comment:
So what happens to data that is larger than 32 pages long? Given our default 256 rows (I think)  that is 8K rows.

Reply:
Correct, but of course the agent can request a different page size.  Effectively making it unlimited by default causes a large initial delay or delay with the second page.

Reply:
I am confused, is total results are truncated to 32 pages, or first fetch is 32 pages long?

Reply:
Fetch 32 pages at a time.   Once you need page 33, you fetch the next 32 pages.  Rather than attempting to fetch/cache the entire result set up front.  That works well until the size starts to become too large.

Reply:
So, on page 32, what is the next link you set? did I miss something here?

Reply:
All of the mechanics are still in place as before it's just that you are limiting how much of the result set you work with at one time.  That's why there are adjustments to the count/skip so that you work correctly over each window.